src/nspawn/nspawn-mount.c

```diff
@@ -426,10 +426,17 @@ int mount_sysfs(const char *dest, MountSettingsMask mount_settings) {
         if (rmdir(full) < 0)
                 return log_error_errno(errno, "Failed to remove %s: %m", full);
 
-        /* Create mountpoint for cgroups. Otherwise we are not allowed since we
-         * remount /sys read-only.
-         */
-        if (cg_ns_supported()) {
+        /* We need to ensure that /sys/fs/cgroup exists before we remount /sys read-only.
+         *
+         * If !use_cgns, then this was already done by the outer child; so we only need to do it here it if use_cgns.
+         * This function doesn't know whether use_cgns, but !cg_ns_supported()⇒!use_cgns, so we can "optimize" the case
+         * where we _know_ !use_cgns, and deal with a no-op mkdir_p() in the false-positive where cgns_supported() but
+         * !use_cgns.
+         *
+         * But is it really much of an optimization?  We're potentially spending an access(2) (cg_ns_supported() could
+         * be cached from a previous call) to potentially save an lstat(2) and mkdir(2); and all of them are on virtual
+         * fileystems, so they should all be pretty cheap. */
```

Comment:
quite franky i'd just simplify this and drop the conditionalization, and just do the mkdir in all cases...

Reply:
/cc @brauner You're the author of that optimization.